sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -253,6 +253,13 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
     dir
   }
 
+  // HIVE-14259 removed FileUtils.isSubDir(). Adapted it from Hive 1.2's FileUtils.isSubDir().
```

Comment:
Hi, @wangyum , @felixcheung , @srowen , @gatorsmile , @cloud-fan , @dbtsai , @rxin .

In general, I have no objection on this PR's approach. But, I want to point out that we are embedding old Hive bug into Spark source code repository logically. This is one of the typical example I want to avoid. HIVE-14259 purposely replaced `public isSubDir` to `private isSubDir` because of the old bug which path `/dir12` is considered as a subdirectory of `/dir1`. Recent Apache Hive releases don't have this bug, only Spark will have now by Spark code. After this PR, I'm worrying about the situation where the whole responsibility falls into Apache Spark community.

Reply:
Can we 'port' the fix here? I agree it's not great to maintain a copy of code, but if a few of these buy us less coupling to Hive and enables an upgrade, it's probably worth it.

Reply:
Any comment, @wangyum ?

Reply:
+1 what Sean says

Reply:
It seems that the fix does not affect Spark:
```scala
  test("hive.exec.stagingdir start with table dir") {
    // hive.exec.stagingdir: file:/private/var/folders/warehouse-c49050a7/test_table_prefix
    // table dir           : file:/private/var/folders/warehouse-c49050a7/test_table
    withSQLConf("hive.exec.stagingdir" -> "../test_table_prefix") {
      withTable("test_table") {
        sql("CREATE TABLE test_table (key int)")
        sql("INSERT OVERWRITE TABLE test_table SELECT 1")
        checkAnswer(sql("SELECT * FROM test_table"), Row(1))
      }
    }
  }
```
But I'd like fix `isSubDir` itself:
```scala
  // HIVE-14259 removed FileUtils.isSubDir(). Adapted it from Hive 1.2's FileUtils.isSubDir().
  private def isSubDir(p1: Path, p2: Path, fs: FileSystem): Boolean = {
    val path1 = fs.makeQualified(p1).toString + Path.SEPARATOR
    val path2 = fs.makeQualified(p2).toString + Path.SEPARATOR
    path1.startsWith(path2)
  }
```